controllers/workspace/finalize.go

```diff
@@ -0,0 +1,202 @@
+//
+// Copyright (c) 2019-2020 Red Hat, Inc.
+// This program and the accompanying materials are made
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"path"
+
+	"github.com/devfile/api/pkg/apis/workspaces/v1alpha2"
+	"github.com/devfile/devworkspace-operator/internal/images"
+	"github.com/devfile/devworkspace-operator/pkg/common"
+	"github.com/devfile/devworkspace-operator/pkg/config"
+	"github.com/go-logr/logr"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	pvcClaimMountPath   = "/tmp/devworkspaces/"
+	cleanupCommandFmt   = "rm -rf %s"
+	pvcCleanupFinalizer = "storage.controller.devfile.io"
+)
+
+var (
+	cleanupJobCompletions  = int32(1)
+	cleanupJobBackoffLimit = int32(3)
+)
+
+func (r *DevWorkspaceReconciler) setFinalizer(ctx context.Context, workspace *v1alpha2.DevWorkspace) (ok bool, err error) {
+	if !isFinalizerNecessary(workspace) || hasFinalizer(workspace) {
+		return true, nil
+	}
+	workspace.SetFinalizers(append(workspace.Finalizers, pvcCleanupFinalizer))
+	return false, r.Update(ctx, workspace)
+}
+
+func (r *DevWorkspaceReconciler) finalize(ctx context.Context, log logr.Logger, workspace *v1alpha2.DevWorkspace) error {
+	if !hasFinalizer(workspace) {
+		return nil
+	}
+	specJob, err := r.getSpecCleanupJob(workspace)
+	if err != nil {
+		return err
+	}
+	clusterJob, err := r.getClusterCleanupJob(ctx, workspace)
+	if err != nil {
+		return err
+	}
+	if clusterJob == nil {
+		err := r.Create(ctx, specJob)
+		if err != nil && !k8sErrors.IsAlreadyExists(err) {
+			return err
+		}
+		return nil
+	}
+	if !equality.Semantic.DeepDerivative(specJob.Spec, clusterJob.Spec) {
+		err := r.Patch(ctx, specJob, client.Merge)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	for _, condition := range clusterJob.Status.Conditions {
+		if condition.Status != corev1.ConditionTrue {
+			continue
+		}
+		switch condition.Type {
+		case batchv1.JobComplete:
+			log.Info("PVC clean up job successful; clearing finalizer")
+			clearFinalizer(workspace)
+			return r.Update(ctx, workspace)
+		case batchv1.JobFailed:
+			log.Error(fmt.Errorf("PVC clean up job failed: message %q", condition.Message),
```

Comment:
Well, I suppose it may fail due to PVC mounting issue if there is another DevWorkspace is running (workspace that is removed was stopped).
And I would say we need to wait until PVC is available to be bound to CleanUp Job.

Also, we should not fail or even run Job if PVC is already removed (it can be implemented even now in https://github.com/devfile/devworkspace-operator/pull/230/files#diff-55c79087cb84a6e155de791d4a71a4506191f8c5eae13cc1ca17ba62ebca6bf8R179)

Reply:
In general, PVC cleanup comes with a ton of corner cases; I think we should run the job whenever the finalizer is present (as it is set up currently) since it being set implies that something was stored in the PVC at some point. If the directory does not exist, `rm` will still return a successful exit code. 

We do need to cover edge cases in mounting PVCs -- I'll look into that.

Reply:
I haven't found an easy way of testing RWO volumes, but I'm fairly certain that the only result when another workspace has the PVC mounted will be pods stuck in a pending state until they can grab it. This shouldn't cause the job to fail.

Reply:
>  but I'm fairly certain that the only result when another workspace has the PVC mounted will be pods stuck in a pending state until they can grab it. This shouldn't cause the job to fail.

That's true. can be reproduced when you remove worksspace PVC